### PR TITLE
Fix Gemma3 prompt caching bug, repetition penalty no longer requires reprocessing

### PR DESCRIPTION
**Pull request by mattjcly**
### Overview

- Previously, to apply repetition penalty properly to cached tokens, we were trimming tokens off the cache and re-computing them.

  1. This was inefficient, as we already computed those tokens
  2. This required a cache trim on basically every single request
  
  This became an issue in situations where the cache could not be trimmed for whatever reason, even when the chat history has been contiguous so it shouldn't have to be trimmed. See https://github.com/lmstudio-ai/mlx-engine/issues/145.
  
  That issue (Gemma 3 cache being un-trimmable when not expected) root causes to a different issue in `mlx-lm` where Gemma 3 doesn't respect `max_kv_size` (Issue: https://github.com/ml-explore/mlx-lm/issues/118).
  
  However, we can avoid that issue altogether by not trying to trim the cache so often for repeat penalty.
  
  Instead, maintain state of cached tokens within the repeat penalty processor, so that they can be prepended to the newly processed/generated tokens before repeat penalty is applied. Additionally, only try to trim the cache if the num tokens to trim is > 0.

- Added a test to verify that this approach results in repeat penalty application as before

- Also moved all "text model" tests to "test_text_models", because naturally different features will interplay with each other so it is not practical to box up tests by "feature" (speculative decoding, structured generation, etc...).

Resolves: https://github.com/lmstudio-ai/mlx-engine/issues/145

#### New

https://github.com/user-attachments/assets/101f6f37-be7f-4618-926f-c0917f833fe5

#### Old

https://github.com/user-attachments/assets/9b440447-fc6e-4cac-9d03-8e360c622465

### CODE REVIEW

File: mlx_engine/processors/repetition_penalty_processor.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        Args:
+            tokens: The tokens to be processed.
+            logits: The logits to be processed.
+        """
+        # append historical tokens s.t. repetition penalty accounts tokens that have already been processed in this gen
+        num_tokens_to_prepend_from_history = max(
+            self.repetition_context_size - len(tokens), 0
+        )
+        historical_tokens = mx.array(
+            self.token_history[-num_tokens_to_prepend_from_history:], dtype=mx.int64
```
Comment (neilmehta24) [inline mlx_engine/processors/repetition_penalty_processor.py:38]: I am not sure that `self.token_history[-num_tokens_to_prepend_from_history:]` works as intended when `num_tokens_to_prepend_from_history` is `0`. I remember having to do this in the past: https://github.com/lmstudio-ai/mlx-engine/blob/8999237a482f185e948b4468b7f2e1d99f8e2964/mlx_engine/processors/outlines_logits_processor.py#L21-L24

File: mlx_engine/processors/repetition_penalty_processor.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    Replace the default repetition penalty processor in logits_processors
+    with a custom one that takes into account the tokens that have already been cached.
+
+    Raises a ValueError if the repetition penalty processor is not found in the logits processors
+    when expected.
+
+    Returns True if the repetition penalty processor was found and replaced, False otherwise.
+    """
+    # exclude the last stream_generate_input tokens from the prompt to get already computed
+    already_computed_tokens = prompt_tokens[:-num_uncached_tokens]
```
Comment (neilmehta24) [inline mlx_engine/processors/repetition_penalty_processor.py:61]: Can `num_uncached_tokens` be `0`?

File: mlx_engine/processors/repetition_penalty_processor.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    # Find and wrap the repetition penalty processor
+    replaced_default_repetition_penalty_processor = False
+    for i, processor in enumerate(logits_processors):
+        processor_name = (
+            processor.__name__ if hasattr(processor, "__name__") else str(processor)
```
Comment (neilmehta24) [inline mlx_engine/processors/repetition_penalty_processor.py:66]: If we're looking for one thing only:
```suggestion
            processor.__name__ if hasattr(processor, "__name__") else None
```

File: mlx_engine/processors/repetition_penalty_processor.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    prompt_tokens: list[int],
+    num_uncached_tokens: int,
+    repetition_context_size: int,
+) -> bool:
+    """
+    Replace the default repetition penalty processor in logits_processors
+    with a custom one that takes into account the tokens that have already been cached.
+
+    Raises a ValueError if the repetition penalty processor is not found in the logits processors
+    when expected.
```
Comment (neilmehta24) [inline mlx_engine/processors/repetition_penalty_processor.py:56]: This doesn't seem to raise a ValueError here

File: tests/test_text_models.py [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+            print("\n", flush=True)
+
+        generate()
+        self.assertGreater(len(generated_text), 0, "Model failed to generate any text")
+        self.assertNotIn(
+            "The quick brown fox jumped over the lazy dog.", generated_text
+        )
+
+    def test_prompt_caching(self):
+        model_path = model_getter("mlx-community/gemma-3-text-4b-it-4bit")
```
Comment (neilmehta24) [inline tests/test_text_models.py:57]: Can this be the `1b` model?

File: tests/test_text_models.py [hunk 6]
```diff
@@ -0,0 +1,409 @@
...
+        prompt_tokens = tokenize(model_kit, prompt)
+        log_info(
+            prefix="test_prompt_caching_trim",
+            message=f"Generation 2 number of prompt tokens: {len(prompt_tokens)}",
+        )
+        generated_text_list_2 = []
+        prompt_progress_callback_times_called = 0
+        generate(text_accumulator=generated_text_list_2)
+        generated_text_2 = "".join(generated_text_list_2)
+        # Expect prompt cache ot be intact for the first half of the file_content, so we should get 1
```
Comment (neilmehta24) [inline tests/test_text_models.py:205]:
```suggestion
        # Expect prompt cache to be intact for the first half of the file_content, so we should get 1
```

File: tests/test_text_models.py [hunk 7]
```diff
@@ -0,0 +1,409 @@
...
+        self.assertNotIn(
+            "The quick brown fox jumped over the lazy dog.", generated_text
+        )
+
+    def test_prompt_caching(self):
+        model_path = model_getter("mlx-community/gemma-3-text-4b-it-4bit")
+        model_kit = load_model(model_path=model_path, max_kv_size=4096)
+        file_content = read_text_file(
+            self.test_data_dir / "ben_franklin_autobiography_start.txt"
+        )
```
Comment (neilmehta24) [inline tests/test_text_models.py:61]:
```suggestion
        file_path = self.test_data_dir / "ben_franklin_autobiography_start.txt"
        file_content = file_path.read_text()
```
and can probably get rid of `read_text_file`

File: mlx_engine/generate.py [hunk 8]
```diff
@@ -214,11 +217,25 @@ def create_generator(
...
+        prompt_tokens,
+        num_uncached_tokens=len(stream_generate_input),
+        repetition_context_size=repetition_context_size,
+    )
+    # If we don't find the repetition penalty processor as expected, there was an mlx-lm change to the processor name
+    # that we did not catch. This error should never be thrown in production since tests should catch
+    if not replace_successful and repetition_penalty is not None:
+        raise ValueError(
+            "Repetition penalty processor unable to be found and replaced, even though repetition_penalty is set"
+        )
```
Comment (neilmehta24) [inline mlx_engine/generate.py:237]: Can we replace this with a warning log, and add a test that checks for this warning log? Can check for the warning with [redirect_stderr](https://docs.python.org/3/library/contextlib.html#contextlib.redirect_stderr) . Not sure if completely disabling the engine is the right move here, because tests can fail to run or be skipped on occasion due to infrastructure/process/human errors

Comment (mattjcly) [reply mlx_engine/generate.py:237]: I'm going to make the change to use our own repetition penalty logits processor completely, and refer to the mlx-lm implementation. Let me know if you think this isn't a better resolution for any reason

Comment (neilmehta24) [reply mlx_engine/generate.py:237]: I am fine with re-implementing the repetition processor. Whatever it takes to avoid completely bricking our MLX impl

File: tests/test_text_models.py (hunk 7 above)
Comment (mattjcly) [reply tests/test_text_models.py:61]: Nice, didn't know about this

File: tests/test_text_models.py (hunk 5 above)
Comment (mattjcly) [reply tests/test_text_models.py:57]: Somehow missed the existence of the 1b model. Yes

Comment (mattjcly) [reply tests/test_text_models.py:57]: Actually seems like https://huggingface.co/mlx-community/gemma-3-1b-it-4bit/tree/main is different in structure from https://huggingface.co/mlx-community/gemma-3-text-4b-it-4bit. Gonna see if the failure case pre-this pr still holds for that model and is fixed by this change, in which case will use

Comment (mattjcly) [reply tests/test_text_models.py:57]: Just switched to qwen 0.5B that we already use

File: mlx_engine/processors/repetition_penalty_processor.py (hunk 4 above)
Comment (mattjcly) [reply mlx_engine/processors/repetition_penalty_processor.py:56]: removed

File: mlx_engine/processors/repetition_penalty_processor.py (hunk 3 above)
Comment (mattjcly) [reply mlx_engine/processors/repetition_penalty_processor.py:66]: removed this part

File: mlx_engine/processors/repetition_penalty_processor.py (hunk 1 above)
Comment (mattjcly) [reply mlx_engine/processors/repetition_penalty_processor.py:38]: Good callout:
Screenshot 2025-04-23 at 5 24 51 PM

Made fix

File: mlx_engine/processors/repetition_penalty_processor.py (hunk 2 above)
Comment (mattjcly) [reply mlx_engine/processors/repetition_penalty_processor.py:61]: good callout, also fixed

Comment (mattjcly) [reply mlx_engine/processors/repetition_penalty_processor.py:61]: (this one should never be 0)

File: mlx_engine/logging.py [hunk 9]
```diff
@@ -1,3 +1,19 @@
+"""
```
[the remark this answers is not in the document]
Comment (mattjcly) [reply mlx_engine/logging.py:1]: Will differ for now, what we have here works find for what we need atm

File: mlx_engine/generate.py (hunk 8 above)
Comment (mattjcly) [reply mlx_engine/generate.py]: Was actually able to just call `make_repetition_penalty` from within the `RepetitionPenaltyProcessor`, so we're not diverging at all. And we never used logits processor anyway so I think this is in fact a cleaner approach

File: mlx_engine/generate.py [hunk 10]
```diff
@@ -221,19 +217,18 @@ def create_generator(
...
-    )
-    # If we don't find the repetition penalty processor as expected, there was an mlx-lm change to the processor name
-    # that we did not catch. This error should never be thrown in production since tests should catch
-    if not replace_successful and repetition_penalty is not None:
-        raise ValueError(
-            "Repetition penalty processor unable to be found and replaced, even though repetition_penalty is set"
+    # Setup logits processors
+    generate_args["logits_processors"] = []
+    if repetition_penalty and repetition_penalty != 0.0:
+        already_computed_tokens = (
```
Comment (neilmehta24) [inline mlx_engine/generate.py:223]: `cached_tokens` instead of `already_computed_tokens`? `computed` I think is a bit vague here

File: mlx_engine/processors/repetition_penalty_processor.py [hunk 11]
```diff
@@ -13,13 +13,15 @@
 class RepetitionPenaltyProcessor:
     def __init__(
         self,
-        original_repetition_penalty_processor: Callable[[mx.array, mx.array], mx.array],
         token_history: list[int],
-        repetition_context_size: int = None,
+        repetition_penalty: float,
+        repetition_context_size: int = 20,
```
Comment (neilmehta24) [inline mlx_engine/processors/repetition_penalty_processor.py:18]: `repetition_context_size: int`

The `repetition_context_size` arg from `create_generator` should be the source of truth for the default

**Review by neilmehta24 — APPROVED** [review]
Approved pending comments
